relays/lib-substrate-relay/src/messages_source.rs

```diff
@@ -571,71 +592,46 @@ where
 	.unwrap_or(Err(SubstrateError::BridgePalletIsNotInitialized))
 }
 
-fn make_message_details_map<C: Chain>(
-	weights: Vec<bp_messages::OutboundMessageDetails<C::Balance>>,
+fn validate_out_msgs_details<C: Chain>(
+	out_msgs_details: &[OutboundMessageDetails<C::Balance>],
 	nonces: RangeInclusive<MessageNonce>,
-) -> Result<MessageDetailsMap<C::Balance>, SubstrateError> {
+) -> Result<(), SubstrateError> {
 	let make_missing_nonce_error = |expected_nonce| {
 		Err(SubstrateError::Custom(format!(
 			"Missing nonce {} in message_details call result. Expected all nonces from {:?}",
 			expected_nonce, nonces,
 		)))
 	};
 
-	let mut weights_map = MessageDetailsMap::new();
-
-	// this is actually prevented by external logic
-	if nonces.is_empty() {
-		return Ok(weights_map)
+	let mut nonces_iter = nonces.clone().rev().peekable();
+	let mut out_msgs_details_iter = out_msgs_details.iter().rev();
+	while let Some((out_msg_details, &nonce)) = out_msgs_details_iter.next().zip(nonces_iter.peek())
+	{
+		nonces_iter.next();
+		if out_msg_details.nonce != nonce {
+			// Some nonces are missing from the middle/tail of the range. This is critical error.
+			return make_missing_nonce_error(nonce)
+		}
 	}
 
-	// check if last nonce is missing - loop below is not checking this
-	let last_nonce_is_missing =
-		weights.last().map(|details| details.nonce != *nonces.end()).unwrap_or(true);
-	if last_nonce_is_missing {
+	// Check if last nonce is missing. The loop above is not checking this.
+	if !nonces.is_empty() && nonces_iter.peek() == Some(nonces.end()) {
```

Comment:
Please verify me here: `nonces_iter` is `noces.rev()`, so we're moving from end to begin and here we are comparing with the `end()` here. Is that intended?

Reply:
Yes, if `nonces_iter.peek() == Some(nonces.end())` at this point, it means that the last nonce was not found in the messages list (and that the messages list is empty actually), and we need to return an error.

Reply:
Ok, can we then check "that the messages list is empty"? :) Imo it should be simpler :)

Reply:
Yes, I think it should work.

Reply:
Done. Also added another check for the case where there are more messages than nonces.